File: utils/shopify-client.py
"""
Shopify Client Utility (Python)
Helper class for custom tooling to interact with Shopify using stored credentials
"""

import boto3
import requests
from typing import Dict, Any, Optional


class ShopifyClient:
    """Client for interacting with Shopify API using credentials from Parameter Store"""

    def __init__(self, shop: str, region: str = 'eu-central-1', api_version: str = '2024-01'):
        self.shop = shop
        self.region = region
        self.api_version = api_version
        self.ssm = boto3.client('ssm', region_name=region)
        self._access_token: Optional[str] = None

    def get_access_token(self) -> str:
        """Get access token from Parameter Store"""
        if self._access_token:
            return self._access_token

        response = self.ssm.get_parameter(
            Name=f'/shopify/clients/{self.shop}/access-token',
            WithDecryption=True
        )
        self._access_token = response['Parameter']['Value']
        return self._access_token

    def graphql(self, query: str, variables: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        """Make a GraphQL API request"""
        token = self.get_access_token()
        url = f'https://{self.shop}/admin/api/{self.api_version}/graphql.json'

        headers = {
            'X-Shopify-Access-Token': token,
            'Content-Type': 'application/json',
        }

        payload = {'query': query}
        if variables:
            payload['variables'] = variables

        response = requests.post(url, json=payload, headers=headers)
        response.raise_for_status()

        data = response.json()

        if 'errors' in data:
            raise Exception(f"GraphQL errors: {data['errors']}")

        return data.get('data', {})

    def request(
        self,
        endpoint: str,
        method: str = 'GET',
        data: Optional[Dict[str, Any]] = None
    ) -> Dict[str, Any]:
        """Make a REST API request"""
        token = self.get_access_token()
        url = f'https://{self.shop}/admin/api/{self.api_version}/{endpoint}'

        headers = {
            'X-Shopify-Access-Token': token,
            'Content-Type': 'application/json',
        }

        response = requests.request(method, url, json=data, headers=headers)
        response.raise_for_status()

        return response.json()

    def get_products(self, limit: int = 50) -> Dict[str, Any]:
        """Get products"""
        return self.request(f'products.json?limit={limit}')

    def get_product(self, product_id: int) -> Dict[str, Any]:
        """Get product by ID"""
        return self.request(f'products/{product_id}.json')

    def get_orders(self, limit: int = 50, status: str = 'any') -> Dict[str, Any]:
        """Get orders"""
        return self.request(f'orders.json?limit={limit}&status={status}')

    def get_order(self, order_id: int) -> Dict[str, Any]:
        """Get order by ID"""
        return self.request(f'orders/{order_id}.json')

    def create_product(self, product_data: Dict[str, Any]) -> Dict[str, Any]:
        """Create a product"""
        return self.request('products.json', method='POST', data={'product': product_data})

    def update_product(self, product_id: int, product_data: Dict[str, Any]) -> Dict[str, Any]:
        """Update a product"""
        return self.request(
            f'products/{product_id}.json',
            method='PUT',
            data={'product': product_data}
        )

    def delete_product(self, product_id: int) -> None:
        """Delete a product"""
        self.request(f'products/{product_id}.json', method='DELETE')


# Example usage:
"""
from shopify_client import ShopifyClient

# Initialize client
client = ShopifyClient('my-store.myshopify.com')

# Get products
products = client.get_products()
print(products)

# GraphQL query
data = client.graphql('''
    {
        products(first: 10) {
            edges {
                node {
                    id
                    title
                    handle
                }
            }
        }
    }
''')
print(data)
"""

File: utils/shopify-client.js
/**
 * Shopify Client Utility
 * Helper functions for custom tooling to interact with Shopify using stored credentials
 */

const { SSMClient, GetParameterCommand } = require('@aws-sdk/client-ssm');

class ShopifyClient {
  constructor(shop, region = 'eu-central-1') {
    this.shop = shop;
    this.region = region;
    this.ssm = new SSMClient({ region });
    this.accessToken = null;
    this.apiVersion = '2024-01';
  }

  /**
   * Get access token from Parameter Store
   */
  async getAccessToken() {
    if (this.accessToken) {
      return this.accessToken;
    }

    const command = new GetParameterCommand({
      Name: `/shopify/clients/${this.shop}/access-token`,
      WithDecryption: true,
    });

    const response = await this.ssm.send(command);
    this.accessToken = response.Parameter.Value;
    return this.accessToken;
  }

  /**
   * Make a GraphQL API request
   */
  async graphql(query, variables = {}) {
    const token = await this.getAccessToken();

    const response = await fetch(`https://${this.shop}/admin/api/${this.apiVersion}/graphql.json`, {
      method: 'POST',
      headers: {
        'X-Shopify-Access-Token': token,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ query, variables }),
    });

    if (!response.ok) {
      throw new Error(`GraphQL request failed: ${response.statusText}`);
    }

    const data = await response.json();

    if (data.errors) {
      throw new Error(`GraphQL errors: ${JSON.stringify(data.errors)}`);
    }

    return data.data;
  }

  /**
   * Make a REST API request
   */
  async request(endpoint, options = {}) {
    const token = await this.getAccessToken();

    const url = `https://${this.shop}/admin/api/${this.apiVersion}/${endpoint}`;

    const response = await fetch(url, {
      ...options,
      headers: {
        'X-Shopify-Access-Token': token,
        'Content-Type': 'application/json',
        ...options.headers,
      },
    });

    if (!response.ok) {
      throw new Error(`API request failed: ${response.statusText}`);
    }

    return await response.json();
  }

  /**
   * Get products
   */
  async getProducts(limit = 50) {
    return await this.request(`products.json?limit=${limit}`);
  }

  /**
   * Get product by ID
   */
  async getProduct(productId) {
    return await this.request(`products/${productId}.json`);
  }

  /**
   * Get orders
   */
  async getOrders(limit = 50, status = 'any') {
    return await this.request(`orders.json?limit=${limit}&status=${status}`);
  }

  /**
   * Get order by ID
   */
  async getOrder(orderId) {
    return await this.request(`orders/${orderId}.json`);
  }

  /**
   * Create a product
   */
  async createProduct(productData) {
    return await this.request('products.json', {
      method: 'POST',
      body: JSON.stringify({ product: productData }),
    });
  }

  /**
   * Update a product
   */
  async updateProduct(productId, productData) {
    return await this.request(`products/${productId}.json`, {
      method: 'PUT',
      body: JSON.stringify({ product: productData }),
    });
  }

  /**
   * Delete a product
   */
  async deleteProduct(productId) {
    return await this.request(`products/${productId}.json`, {
      method: 'DELETE',
    });
  }
}

module.exports = ShopifyClient;

// Example usage:
/*
const ShopifyClient = require('./utils/shopify-client');

async function example() {
  const client = new ShopifyClient('my-store.myshopify.com');

  // Get products
  const products = await client.getProducts();
  console.log(products);

  // GraphQL query
  const data = await client.graphql(`
    {
      products(first: 10) {
        edges {
          node {
            id
            title
            handle
          }
        }
      }
    }
  `);
  console.log(data);
}
*/
